# Wrangle and Analyze Twitter Data

#### Import of the needed libraries:

In [1]:
import requests
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns

## Data Gathering

#### 1. WeRateDogs™ Twitter Archive

As advised I downloaded the twitter_archive_enhanced.csv.

In [2]:
df_twitter = pd.read_csv("twitter-archive-enhanced.csv")

df_twitter.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

#### 2. Tweet image predictions

As advised I programmatically downloaded the image_predictions.tsv.

In [3]:
url = "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"

response = requests.get(url)

with open("image_predictions.tsv", mode = "wb") as file:
    file.write(response.content)

335079

In [4]:
df_predict = pd.read_csv("image_predictions.tsv", sep='\t')

df_predict.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

#### 3. Additional Twitter data

I didn't create a dev account but downloaded the file manually.

Testing Tweepy - but it runs forever, so I had to manually stop the kernel here.

In [5]:
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer

# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = df_twitter.tweet_id.values
len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)


2356

1: 892420643555336193
Fail
2: 892177421306343426
Fail
3: 891815181378084864
Fail
4: 891689557279858688
Fail
5: 891327558926688256
Fail
6: 891087950875897856
Fail
7: 890971913173991426
Fail
8: 890729181411237888
Fail
9: 890609185150312448
Fail
10: 890240255349198849
Fail
11: 890006608113172480
Fail
12: 889880896479866881
Fail
13: 889665388333682689
Fail
14: 889638837579907072
Fail
15: 889531135344209921
Fail
16: 889278841981685760
Fail
17: 888917238123831296
Fail
18: 888804989199671297
Fail
19: 888554962724278272
Fail
20: 888202515573088257
Fail
21: 888078434458587136
Fail
22: 887705289381826560
Fail
23: 887517139158093824
Fail
24: 887473957103951883
Fail
25: 887343217045368832
Fail
26: 887101392804085760
Fail
27: 886983233522544640
Fail
28: 886736880519319552
Fail
29: 886680336477933568
Fail
30: 886366144734445568
Fail
31: 886267009285017600
Fail
32: 886258384151887873
Fail
33: 886054160059072513
Fail
34: 885984800019947520
Fail
35: 885528943205470208
Fail
36: 885518971528720385
Fail
3

KeyboardInterrupt: 

In [270]:
additional = pd.read_json('tweet-json.txt', lines=True)

In [271]:
columns = ["id", "retweet_count", "favorite_count", "retweeted", "display_text_range"]
df_add = additional[columns]
df_add.head(5)

id  retweet_count  favorite_count  retweeted  \
0  892420643555336193           8853           39467      False   
1  892177421306343426           6514           33819      False   
2  891815181378084864           4328           25461      False   
3  891689557279858688           8964           42908      False   
4  891327558926688256           9774           41048      False   

  display_text_range  
0            [0, 85]  
1           [0, 138]  
2           [0, 121]  
3            [0, 79]  
4           [0, 138]

In [272]:
df_twitter.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [273]:
df_predict.head(1)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   

   img_num                      p1   p1_conf  p1_dog      p2   p2_conf  \
0        1  Welsh_springer_spaniel  0.465074    True  collie  0.156665   

   p2_dog                 p3   p3_conf  p3_dog  
0    True  Shetland_sheepdog  0.061428    True

In [274]:
df_add.head(1)

id  retweet_count  favorite_count  retweeted  \
0  892420643555336193           8853           39467      False   

  display_text_range  
0            [0, 85]

In [275]:
df_add.rename(columns={'id': 'tweet_id'}, inplace=True)

Renaming of the id column for the additional twitter data, as the other data sources used "tweet_id" - trying to be consistent for future work.

<a id='assessing'></a>
## Data Assessing


Looking for data quality problems etc.

#### df_twitter

Let's first look for missing data.

In [276]:
df_twitter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [277]:
df_twitter.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
27    886736880519319552                    NaN                  NaN   
1124  730211855403241472                    NaN                  NaN   
953   751830394383790080                    NaN                  NaN   
876   761004547850530816                    NaN                  NaN   
273   840728873075638272                    NaN                  NaN   
2203  668641109086707712                    NaN                  NaN   
2323  666428276349472768                    NaN                  NaN   
1917  674291837063053312                    NaN                  NaN   
1907  674447403907457024                    NaN                  NaN   
607   798209839306514432                    NaN                  NaN   

                      timestamp  \
27    2017-07-16 23:58:41 +0000   
1124  2016-05-11 01:44:07 +0000   
953   2016-07-09 17:28:29 +0000   
876   2016-08-04 01:03:17 +0000   
273   2017-03-12 00:59:17 +0000   
2203  2015-11-23 04:03:57 +0000   
2323  2015-11-17 01:30:57 +0000   
1917  2015-12-08 18:17:56 +0000   
1907  2015-12-09 04:36:06 +0000   
607   2016-11-14 17:03:50 +0000   

                                                 source  \
27    <a href="http://twitter.com/download/iphone" r...   
1124  <a href="http://twitter.com/download/iphone" r...   
953   <a href="http://twitter.com/download/iphone" r...   
876   <a href="http://twitter.com/download/iphone" r...   
273   <a href="http://twitter.com/download/iphone" r...   
2203  <a href="http://twitter.com/download/iphone" r...   
2323  <a href="http://twitter.com/download/iphone" r...   
1917  <a href="http://twitter.com/download/iphone" r...   
1907  <a href="http://twitter.com/download/iphone" r...   
607   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
27    This is Mingus. He's a wonderful father to his...                  NaN   
1124  This is Ziva. She doesn't know how her collar ...                  NaN   
953   This is Tucker. He's very camera shy. 12/10 wo...                  NaN   
876   This is Bo and Ty. Bo eats paper and Ty felt l...                  NaN   
273   RT @dog_rates: This is Pipsy. He is a fluffbal...         6.671522e+17   
2203  Meet Hanz. He heard some thunder. 10/10 https:...                  NaN   
2323  Here we have an Austrian Pulitzer. Collectors ...                  NaN   
1917  This is Kenny. He just wants to be included in...                  NaN   
1907  This pupper just wants a belly rub. This puppe...                  NaN   
607   This is Cooper. His bow tie was too heavy for ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
27                         NaN                        NaN   
1124                       NaN                        NaN   
953                        NaN                        NaN   
876                        NaN                        NaN   
273               4.196984e+09  2015-11-19 01:27:25 +0000   
2203                       NaN                        NaN   
2323                       NaN                        NaN   
1917                       NaN                        NaN   
1907                       NaN                        NaN   
607                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
27    https://www.gofundme.com/mingusneedsus,https:/...                13   
1124  https://twitter.com/dog_rates/status/730211855...                11   
953   https://twitter.com/dog_rates/status/751830394...                12   
876   https://twitter.com/dog_rates/status/761004547...                11   
273   https://twitter.com/dog_rates/status/667152164...                12   
2203  https://twitter.com/dog_rates/status/668641109...                10   
2323  https://twitter.com/dog_rates/status/666428276...                 7   
1917  https://twit

A lot of missing data in the retweeted and reply columns!
Since we only want original posts with images, we'll drop them later.
Additionally we can see that:
- not all tweets could be classified correctly
- HTML code in the source column
- records/tweets where the name is "None"
- incorrect datatypes for timestamp columns -> should be datetime instead of objects

In [278]:
df_twitter['name'].value_counts()

None       745
a           55
Charlie     12
Cooper      11
Lucy        11
          ... 
Dotsy        1
Banditt      1
Eugene       1
JD           1
Huck         1
Name: name, Length: 957, dtype: int64

Some names are not real names, like "None" or "a".

In [279]:
df_twitter.nunique()
df_twitter.shape[0]

tweet_id                      2356
in_reply_to_status_id           77
in_reply_to_user_id             31
timestamp                     2356
source                           4
text                          2356
retweeted_status_id            181
retweeted_status_user_id        25
retweeted_status_timestamp     181
expanded_urls                 2218
rating_numerator                40
rating_denominator              18
name                           957
doggo                            2
floofer                          2
pupper                           2
puppo                            2
dtype: int64

2356

No duplicate tweets, as the amount of unique tweet id's is identical to the number of records in the dataset.

In [280]:
check_classification = df_twitter.copy()

check_classification[['doggo','floofer','pupper','puppo']].apply(lambda x: True if
    (x[0] == 'None' and 
    x[1] == 'None' and 
    x[2] == 'None' and 
    x[3] =='None') 
    else False, axis = 1).value_counts()

True     1976
False     380
dtype: int64

Dogs could only be correctly classified 16.13% of the time (380/(1976+380)).

Only for 16% of the rows the data is not missing. Now let's take a look at the ratings. By what we have seen so far, it looks like the ratings have always a format of 13/10 or 12/10 and so on. So we would expect a numerator > 10 and denominator = 10.

In [281]:
df_twitter['rating_numerator'].value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

There are a wide range of rating_numerators in this dataset, going as high as 1776! The scale only goes from one to ten though..

In [282]:
df_twitter.query('rating_numerator == "1776"')['expanded_urls']

979    https://twitter.com/dog_rates/status/749981277...
Name: expanded_urls, dtype: object

Checking the tweet this is a legitimate record though.
So ratings above 10 are possible.

In [283]:
df_twitter.query('rating_numerator == "1"').head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
605   798576900688019456                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1869  675153376133427200                    NaN                  NaN   
1940  673716320723169284           6.737159e+17         4.196984e+09   
2038  671550332464455680           6.715449e+17         4.196984e+09   

                      timestamp  \
605   2016-11-15 17:22:24 +0000   
1446  2016-02-08 00:27:39 +0000   
1869  2015-12-11 03:21:23 +0000   
1940  2015-12-07 04:11:02 +0000   
2038  2015-12-01 04:44:10 +0000   

                                                 source  \
605   <a href="http://twitter.com/download/iphone" r...   
1446  <a href="http://twitter.com/download/iphone" r...   
1869  <a href="http://twitter.com/download/iphone" r...   
1940  <a href="http://twitter.com/download/iphone" r...   
2038  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
605   RT @dog_rates: Not familiar with this breed. N...         6.661041e+17   
1446  After reading the comments I may have overesti...                  NaN   
1869  What kind of person sends in a picture without...                  NaN   
1940  The millennials have spoken and we've decided ...                  NaN   
2038  After 22 minutes of careful deliberation this ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
605               4.196984e+09  2015-11-16 04:02:55 +0000   
1446                       NaN                        NaN   
1869                       NaN                        NaN   
1940                       NaN                        NaN   
2038                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
605   https://twitter.com/dog_rates/status/666104133...                 1   
1446                                                NaN                 1   
1869  https://twitter.com/dog_rates/status/675153376...                 1   
1940                                                NaN                 1   
2038                                                NaN                 1   

      rating_denominator  name doggo floofer pupper puppo  
605                   10  None  None    None   None  None  
1446                  10  None  None    None   None  None  
1869                  10  None  None    None   None  None  
1940                  10  None  None    None   None  None  
2038                  10  None  None    None   None  None

Checking some of the entries we can see that those are joke tweets (a chicken for example).

In [284]:
df_twitter.query('rating_numerator == "0"').head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
315   835152434251116546                    NaN                  NaN   
1016  746906459439529985           7.468859e+17         4.196984e+09   

                      timestamp  \
315   2017-02-24 15:40:31 +0000   
1016  2016-06-26 03:22:31 +0000   

                                                 source  \
315   <a href="http://twitter.com/download/iphone" r...   
1016  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
315   When you're so blinded by your systematic plag...                  NaN   
1016  PUPDATE: can't see any. Even if I could, I cou...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
315                        NaN                        NaN   
1016                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
315   https://twitter.com/dog_rates/status/835152434...                 0   
1016  https://twitter.com/dog_rates/status/746906459...                 0   

      rating_denominator  name doggo floofer pupper puppo  
315                   10  None  None    None   None  None  
1016                  10  None  None    None   None  None

One tweet was deleted, the other was a screenshot of another tweet with a dog in it. Not a very clear picture though.

In [285]:
df_twitter['rating_denominator'].value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [286]:
df_twitter.query('rating_denominator == "170"')['expanded_urls'].values

array(['https://twitter.com/dog_rates/status/731156023742988288/photo/1'],
      dtype=object)

In [287]:
df_twitter.query('rating_denominator == "0"')['expanded_urls'].values
df_twitter.query('rating_denominator == "7"')['expanded_urls'].values

array([nan], dtype=object)

array(['https://www.gofundme.com/sams-smile,https://twitter.com/dog_rates/status/810984652412424192/photo/1'],
      dtype=object)

In [288]:
df_twitter.query('rating_denominator == "0"').head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640           8.352460e+17           26259576.0   

                     timestamp  \
313  2017-02-24 21:54:03 +0000   

                                                source  \
313  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
313  @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp expanded_urls  \
313                       NaN                        NaN           NaN   

     rating_numerator  rating_denominator  name doggo floofer pupper puppo  
313               960                   0  None  None    None   None  None

In [289]:

df_twitter.query('rating_denominator == "7"').head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
516  810984652412424192                    NaN                  NaN   

                     timestamp  \
516  2016-12-19 23:06:23 +0000   

                                                source  \
516  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
516  Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
516                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
516  https://www.gofundme.com/sams-smile,https://tw...                24   

     rating_denominator name doggo floofer pupper puppo  
516                   7  Sam  None    None   None  None

The same problems occur here. Jokes or unclear pictures (multiple dogs in a picture for example). Another problem are wrongly interpreted "/" signs. which we'll fix with RegEx now.

In [290]:
pattern = "(\d+(\.\d+)?\/\d+(\.\d+)?)"

check_classification["rating"] = check_classification['text'].str.extract(pattern, expand = True)[0]

check_classification[['num', 'denom']] = check_classification['rating'].str.split('/', n=1, expand=True)

check_classification['rating_numerator'] = check_classification['rating_numerator'].astype("str")
check_classification['rating_denominator'] = check_classification['rating_denominator'].astype("str")

# check if we extracted the numerators and denominators correctly
check_classification["check_num"] = check_classification[["rating_numerator", "num"]].apply(lambda x: False if (x[0] != x[1]) else True, axis = 1)

check_classification["check_denom"] = check_classification[["rating_denominator", "denom"]].apply(lambda x: False if (x[0] != x[1]) else True, axis = 1)

In [291]:
check_classification["check_num"].value_counts()

True     2349
False       7
Name: check_num, dtype: int64

In [292]:
check_classification.query("check_num == False")[["rating_numerator", "num","check_num"]]

rating_numerator    num  check_num
45                  5   13.5      False
340                75   9.75      False
387                 7    007      False
695                75   9.75      False
763                27  11.27      False
1689                5    9.5      False
1712               26  11.26      False

Those nums somehow got transformed into floats, which is wrong.

In [293]:
check_classification[["rating_numerator", "num", "check_num"]].head(5)

rating_numerator num  check_num
0               13  13       True
1               13  13       True
2               12  12       True
3               13  13       True
4               12  12       True

In [294]:
check_classification[["rating_numerator", "num", "check_num"]].tail(5)

rating_numerator num  check_num
2351                5   5       True
2352                6   6       True
2353                9   9       True
2354                7   7       True
2355                8   8       True

In [295]:
check_classification['check_denom'].value_counts()

True     2355
False       1
Name: check_denom, dtype: int64

In [296]:
check_classification.query("check_denom == False")[["rating_denominator", "denom","check_denom"]]

rating_denominator denom  check_denom
313                  0    00        False

Looks like a problem we can ignore for now.

In [297]:
check_classification[["rating_denominator", "denom", "check_denom"]].head(5)

rating_denominator denom  check_denom
0                 10    10         True
1                 10    10         True
2                 10    10         True
3                 10    10         True
4                 10    10         True

In [298]:
check_classification[["rating_denominator", "denom", "check_denom"]].tail(5)

rating_denominator denom  check_denom
2351                 10    10         True
2352                 10    10         True
2353                 10    10         True
2354                 10    10         True
2355                 10    10         True

Looks good now!

In [299]:
check_classification["count"] = check_classification['text'].str.count(pattern)

In [300]:
check_classification["count"].value_counts()

1    2323
2      32
3       1
Name: count, dtype: int64

In [301]:
pd.set_option('display.max_columns', 500)
pd.set_option('display.max_colwidth', 4000)

check_classification[['text', 'count']].query('count != 1').head(10)

text  \
55                                                     @roushfenway These are good dogs but 17/10 is an emotional impulse rating. More like 13/10s   
313                                             @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho   
561             RT @dog_rates: "Yep... just as I suspected. You're not flossing." 12/10 and 11/10 for the pup not flossing https://t.co/SuXcI9B7pQ   
766                            "Yep... just as I suspected. You're not flossing." 12/10 and 11/10 for the pup not flossing https://t.co/SuXcI9B7pQ   
784   RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…   
860   RT @dog_rates: Meet Eve. She's a raging alcoholic 8/10 (would b 11/10 but pupper alcoholism is a tragic issue that I can't condone) https:/…   
1007              This is Bookstore and Seaweed. Bookstore is tired and Seaweed is an asshole. 10/10 and 7/10 respectively https://t.co/eUGjGjjFVJ   
1068  After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   
1165                                                                              Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a   
1202                         This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq   

      count  
55        2  
313       2  
561       2  
766       2  
784       2  
860       2  
1007      2  
1068      2  
1165      2  
1202      2

As we can see there are tweets with multiple ratings in it.

This includes retweets!

And some of those are ratings which do not conform to the one-to-ten-rating scale.

#### df_predict

Let's also look for missing data first.

In [302]:
df_predict.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [303]:
df_predict.isna().sum()

tweet_id    0
jpg_url     0
img_num     0
p1          0
p1_conf     0
p1_dog      0
p2          0
p2_conf     0
p2_dog      0
p3          0
p3_conf     0
p3_dog      0
dtype: int64

In [304]:
df_predict.head(10)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5  666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6  666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7  666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8  666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9  666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   
5        1    Bernese_mountain_dog  0.651137    True    English_springer   
6        1              box_turtle  0.933012   False          mud_turtle   
7        1                    chow  0.692517    True     Tibetan_mastiff   
8        1           shopping_cart  0.962465   False     shopping_basket   
9        1        miniature_poodle  0.201493    True            komondor   

    p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0  0.156665    True            Shetland_sheepdog  0.061428    True  
1  0.074192    True          Rhodesian_ridgeback  0.072010    True  
2  0.138584    True                   bloodhound  0.116197    True  
3  0.360687    True           miniature_pinscher  0.222752    True  
4  0.243682    True                     Doberman  0.154629    True  
5  0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True  
6  0.045885   False                     terrapin  0.017885   False  
7  0.058279    True                     fur_coat  0.054449   False  
8  0.014594   False             golden_retriever  0.007959    True  
9  0.192305    True  soft-coated_wheaten_terrier  0.082086    True

No null values at all! But we can clearly see that there is no standardized format for the dog breed/race prediction as they sometimes start with capital letters and sometimes they do not (German_shepherd but miniature_pinscher).

Additionally they are separated by and underscore instead of a whitespace.

Lastly there are pictures, which have no prediction at all (p1_dog, p2_dog and p3_dog are all False).

In [305]:
df_predict['jpg_url'].value_counts().head(10)

https://pbs.twimg.com/media/CcG07BYW0AErrC9.jpg                                            2
https://pbs.twimg.com/media/Ck2d7tJWUAEPTL3.jpg                                            2
https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg                                            2
https://pbs.twimg.com/media/Co-hmcYXYAASkiG.jpg                                            2
https://pbs.twimg.com/media/CwJR1okWIAA6XMp.jpg                                            2
https://pbs.twimg.com/media/CsVO7ljW8AAckRD.jpg                                            2
https://pbs.twimg.com/ext_tw_video_thumb/807106774843039744/pu/img/8XZg1xW35Xp2J6JW.jpg    2
https://pbs.twimg.com/media/CuRDF-XWcAIZSer.jpg                                            2
https://pbs.twimg.com/media/CrXhIqBW8AA6Bse.jpg                                            2
https://pbs.twimg.com/media/CsGnz64WYAEIDHJ.jpg                                            2
Name: jpg_url, dtype: int64

As we can't compare the pictures we'll use the jpg_url to identify duplicates. There are some with multiple entries, meaning that those must be retweets or people just posting multiple pictures of the same dog.

In [306]:
df_predict[df_predict['tweet_id'].duplicated()]

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

In [307]:
df_predict[df_predict['jpg_url'] == 'https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg']

tweet_id                                          jpg_url  \
713   685325112850124800  https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg   
1368  761750502866649088  https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg   

      img_num                p1   p1_conf  p1_dog                  p2  \
713         1  golden_retriever  0.586937    True  Labrador_retriever   
1368        1  golden_retriever  0.586937    True  Labrador_retriever   

      p2_conf  p2_dog      p3  p3_conf  p3_dog  
713   0.39826    True  kuvasz  0.00541    True  
1368  0.39826    True  kuvasz  0.00541    True

In [308]:
df_twitter.query('tweet_id == 798697898615730177')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
596  798697898615730177                    NaN                  NaN   

                     timestamp  \
596  2016-11-16 01:23:12 +0000   

                                                                                 source  \
596  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                               text  \
596  RT @dog_rates: This is Stubert. He just arrived. 10/10 https://t.co/HVGs5aAKAn   

     retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
596         7.128090e+17              4.196984e+09  2016-03-24 01:11:29 +0000   

                                                                                                                       expanded_urls  \
596  https://twitter.com/dog_rates/status/712809025985978368/photo/1,https://twitter.com/dog_rates/status/712809025985978368/photo/1   

     rating_numerator  rating_denominator     name doggo floofer pupper puppo  
596                10                  10  Stubert  None    None   None  None

Confirmed retweets as the tweet_id is different.

We only want Tweets with pictures which contain dogs. Let's see if there are pictures, for which the ML - Algorithm didn't predict any dogs.

In [309]:
df_predict.query('p1_dog == False and p2_dog == False and p3_dog == False').head(10)

tweet_id                                          jpg_url  \
6   666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
17  666104133288665088  https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18  666268910803644416  https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21  666293911632134144  https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
25  666362758909284353  https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
29  666411507551481857  https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
45  666786068205871104  https://pbs.twimg.com/media/CUDmZIkWcAAIPPe.jpg   
50  666837028449972224  https://pbs.twimg.com/media/CUEUva1WsAA2jPb.jpg   
51  666983947667116034  https://pbs.twimg.com/media/CUGaXDhW4AY9JUH.jpg   
53  667012601033924608  https://pbs.twimg.com/media/CUG0bC0U8AAw2su.jpg   

    img_num                p1   p1_conf  p1_dog                   p2  \
6         1        box_turtle  0.933012   False           mud_turtle   
17        1               hen  0.965932   False                 cock   
18        1  desktop_computer  0.086502   False                 desk   
21        1  three-toed_sloth  0.914671   False                otter   
25        1        guinea_pig  0.996496   False                skunk   
29        1              coho  0.404640   False           barracouta   
45        1             snail  0.999888   False                 slug   
50        1       triceratops  0.442113   False            armadillo   
51        1              swab  0.589446   False            chain_saw   
53        1             hyena  0.987230   False  African_hunting_dog   

     p2_conf  p2_dog              p3   p3_conf  p3_dog  
6   0.045885   False        terrapin  0.017885   False  
17  0.033919   False       partridge  0.000052   False  
18  0.085547   False        bookcase  0.079480   False  
21  0.015250   False  great_grey_owl  0.013207   False  
25  0.002402   False         hamster  0.000461   False  
29  0.271485   False             gar  0.189945   False  
45  0.000055   False           acorn  0.000026   False  
50  0.114071   False   common_iguana  0.043255   False  
51  0.190142   False             wig  0.034510   False  
53  0.012601   False          coyote  0.000057   False

In [310]:
df_predict.query('p1_dog == False and (p2_dog == True or p3_dog == True)').head(10)

tweet_id                                          jpg_url  \
8    666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
22   666337882303524864  https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
33   666430724426358785  https://pbs.twimg.com/media/CT-jNYqW4AAPi2M.jpg   
43   666776908487630848  https://pbs.twimg.com/media/CUDeDoWUYAAD-EM.jpg   
52   666996132027977728  https://pbs.twimg.com/media/CUGlb6iUwAITEbW.jpg   
87   667524857454854144  https://pbs.twimg.com/media/CUOGUfJW4AA_eni.jpg   
95   667550904950915073  https://pbs.twimg.com/media/CUOb_gUUkAACXdS.jpg   
103  667806454573760512  https://pbs.twimg.com/media/CUSGbXeVAAAgztZ.jpg   
108  667878741721415682  https://pbs.twimg.com/media/CUTILFiWcAE8Rle.jpg   
113  667915453470232577  https://pbs.twimg.com/media/CUTpj-GWcAATc6A.jpg   

     img_num                  p1   p1_conf  p1_dog                  p2  \
8          1       shopping_cart  0.962465   False     shopping_basket   
22         1                  ox  0.416669   False        Newfoundland   
33         1               llama  0.505184   False       Irish_terrier   
43         1           seat_belt  0.375057   False  miniature_pinscher   
52         1                 hay  0.507637   False          Rottweiler   
87         1                hare  0.447893   False               dhole   
95         1            web_site  0.999335   False              vizsla   
103        1             toyshop  0.253089   False           Chihuahua   
108        1           seat_belt  0.200373   False  miniature_pinscher   
113        1  leatherback_turtle  0.452517   False               boxer   

      p2_conf  p2_dog                        p3   p3_conf  p3_dog  
8    0.014594   False          golden_retriever  0.007959    True  
22   0.278407    True               groenendael  0.102643    True  
33   0.104109    True                     dingo  0.062071   False  
43   0.167175    True                 Chihuahua  0.086951    True  
52   0.062490    True             water_buffalo  0.048425   False  
87   0.092435   False  Chesapeake_Bay_retriever  0.088122    True  
95   0.000081    True                    collie  0.000069    True  
103  0.187155    True         Brabancon_griffon  0.112799    True  
108  0.106003    True                schipperke  0.104733    True  
113  0.196655    True                  terrapin  0.160983   False

After checking some of these pictures it gets clear, that sometimes the dogs are in the background or the pictures don't contain any dogs at all.

#### df_add

Lets repeat the same pattern for this dataset: missing data!

In [311]:
df_add.head(10)

tweet_id  retweet_count  favorite_count  retweeted  \
0  892420643555336193           8853           39467      False   
1  892177421306343426           6514           33819      False   
2  891815181378084864           4328           25461      False   
3  891689557279858688           8964           42908      False   
4  891327558926688256           9774           41048      False   
5  891087950875897856           3261           20562      False   
6  890971913173991426           2158           12041      False   
7  890729181411237888          16716           56848      False   
8  890609185150312448           4429           28226      False   
9  890240255349198849           7711           32467      False   

  display_text_range  
0            [0, 85]  
1           [0, 138]  
2           [0, 121]  
3            [0, 79]  
4           [0, 138]  
5           [0, 138]  
6           [0, 140]  
7           [0, 118]  
8           [0, 122]  
9           [0, 133]

In [312]:
df_add.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2354 non-null   int64 
 1   retweet_count       2354 non-null   int64 
 2   favorite_count      2354 non-null   int64 
 3   retweeted           2354 non-null   bool  
 4   display_text_range  2354 non-null   object
dtypes: bool(1), int64(3), object(1)
memory usage: 76.0+ KB


No missing values here.

<a id='cleaning'></a>
## Data Cleaning

Cleaning steps:

**1. Merge tables - TIDINESS ISSUE!** 

**2. Drop everything we found before. Retweets, replies, tweets without an image or tweets without dogs - QUALITY ISSUE!**

**3. Fixing datatypes  - QUALITY ISSUE!**

**4. Clean the source column - QUALITY ISSUE!**

**5. Split the text range into two separate columns - TIDINESS ISSUE!**

**6. Merging the different classification columns (doggo, floofer etc.) into one column and removing "None" entries - TIDINESS ISSUE!**

**7. Remove incorrect names - QUALITY ISSUE!**

**8. Just use the column with the highest % and the respective prediction - TIDINESS ISSUE!**

**9. Standardize the breeds, because some are uppercase - QUALITY ISSUE!**

**1. Merge tables - TIDINESS ISSUE!**

As the tweets als have a tweet_id and cleaning would be way easier in one big DataFrame, we'll merge those three DataFrames together now.

In [313]:
df_main = pd.merge(df_twitter, df_add, on = 'tweet_id', how = 'outer')
df_main = pd.merge(df_main, df_predict, on = 'tweet_id', how = 'outer')
df_main_cleaned = df_main.copy()

In [314]:
df_main_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 32 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

**2. Drop everything we found before. Retweets, replies, tweets without an image or tweets without dogs - QUALITY ISSUE!**

In [315]:
pd.set_option('display.max_colwidth', 50)
df_main_cleaned[df_main_cleaned['retweeted'].isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19   888202515573088257                    NaN                  NaN   
815  771004394259247104                    NaN                  NaN   

                     timestamp  \
19   2017-07-21 01:02:36 +0000   
815  2016-08-31 15:19:06 +0000   

                                                source  \
19   <a href="http://twitter.com/download/iphone" r...   
815  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
19   RT @dog_rates: This is Canela. She attempted s...         8.874740e+17   
815  RT @katieornah: @dog_rates learning a lot at c...         7.710021e+17   

     retweeted_status_user_id retweeted_status_timestamp  \
19               4.196984e+09  2017-07-19 00:47:34 +0000   
815              1.732729e+09  2016-08-31 15:10:07 +0000   

                                         expanded_urls  rating_numerator  \
19   https://twitter.com/dog_rates/status/887473957...                13   
815  https://twitter.com/katieornah/status/77100213...                12   

     rating_denominator    name doggo floofer  pupper puppo  retweet_count  \
19                   10  Canela  None    None    None  None            NaN   
815                  10    None  None    None  pupper  None            NaN   

     favorite_count retweeted display_text_range  \
19              NaN       NaN                NaN   
815             NaN       NaN                NaN   

                                             jpg_url  img_num            p1  \
19   https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg      2.0      Pembroke   
815  https://pbs.twimg.com/media/CrMmVqyWcAIDCHI.jpg      1.0  home_theater   

      p1_conf p1_dog                   p2   p2_conf p2_dog          p3  \
19   0.809197   True  Rhodesian_ridgeback  0.054950   True      beagle   
815  0.414338  False                 iPod  0.052741  False  pop_bottle   

      p3_conf p3_dog  
19   0.038915   True  
815  0.048821  False

In [316]:
df_main_cleaned = df_main_cleaned[df_main_cleaned['retweeted_status_id'].isnull()]
df_main_cleaned = df_main_cleaned[df_main_cleaned['in_reply_to_status_id'].isnull()]
df_main_cleaned.dropna(subset = ["retweeted"], inplace = True)

In [317]:
df_main_cleaned['retweeted'].value_counts()

False    2097
Name: retweeted, dtype: int64

No more retweets in the dataset and no more records with missing data.

In [318]:
df_main_cleaned.drop(['in_reply_to_status_id', 'in_reply_to_user_id',
                      'retweeted_status_id', 'retweeted_status_user_id', 
                      'retweeted_status_timestamp', 'retweeted'], inplace=True, axis = 1)

We've dropped the unnecessary columns now.

In [319]:
df_main_cleaned.dropna(subset = ['jpg_url'], inplace = True)
df_main_cleaned['jpg_url'].duplicated().sum()

0

Dropped records with a missing image.

In [320]:
# As noted in the comments, this action could delete rows with useful information. Just because the algorithm couldn't predict a dog, doesn't mean that there was no dog in the picture!
# df_main_cleaned.drop(df_main_cleaned.query('p1_dog == False and p2_dog == False and p3_dog == False').index, inplace = True)
# df_main_cleaned.query('p1_dog == False and p2_dog == False and p3_dog == False')

In [321]:
df_main_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1971 entries, 0 to 2355
Data columns (total 26 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tweet_id            1971 non-null   int64  
 1   timestamp           1971 non-null   object 
 2   source              1971 non-null   object 
 3   text                1971 non-null   object 
 4   expanded_urls       1971 non-null   object 
 5   rating_numerator    1971 non-null   int64  
 6   rating_denominator  1971 non-null   int64  
 7   name                1971 non-null   object 
 8   doggo               1971 non-null   object 
 9   floofer             1971 non-null   object 
 10  pupper              1971 non-null   object 
 11  puppo               1971 non-null   object 
 12  retweet_count       1971 non-null   float64
 13  favorite_count      1971 non-null   float64
 14  display_text_range  1971 non-null   object 
 15  jpg_url             1971 non-null   object 
 16  img_nu

Lastly we dropped records which did not contain dogs.

**3. Fixing datatypes - QUALITY ISSUE!**

In [322]:
df_main_cleaned['tweet_id'] = df_main_cleaned['tweet_id'].astype('str')
df_main_cleaned['timestamp'] = pd.to_datetime(df_main_cleaned['timestamp'])
for x in ['retweet_count', 'favorite_count', 'img_num']:
    df_main_cleaned[x] = df_main_cleaned[x].astype('int64')

In [323]:
df_main_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1971 entries, 0 to 2355
Data columns (total 26 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            1971 non-null   object             
 1   timestamp           1971 non-null   datetime64[ns, UTC]
 2   source              1971 non-null   object             
 3   text                1971 non-null   object             
 4   expanded_urls       1971 non-null   object             
 5   rating_numerator    1971 non-null   int64              
 6   rating_denominator  1971 non-null   int64              
 7   name                1971 non-null   object             
 8   doggo               1971 non-null   object             
 9   floofer             1971 non-null   object             
 10  pupper              1971 non-null   object             
 11  puppo               1971 non-null   object             
 12  retweet_count       1971 non-null 

**4. Clean the source column - QUALITY ISSUE!**

The source column has HTML clutter all in it, which makes the column very hard to use. We'll extract only the essential stuff.

In [324]:
df_main_cleaned.tail(5)

tweet_id                 timestamp  \
2351  666049248165822465 2015-11-16 00:24:50+00:00   
2352  666044226329800704 2015-11-16 00:04:52+00:00   
2353  666033412701032449 2015-11-15 23:21:54+00:00   
2354  666029285002620928 2015-11-15 23:05:30+00:00   
2355  666020888022790149 2015-11-15 22:32:08+00:00   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
2351  Here we have a 1949 1st generation vulpix. Enj...   
2352  This is a purebred Piers Morgan. Loves to Netf...   
2353  Here is a very happy pup. Big fan of well-main...   
2354  This is a western brown Mitsubishi terrier. Up...   
2355  Here we have a Japanese Irish Setter. Lost eye...   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  retweet_count  \
2351                  10  None  None    None   None  None             41   
2352                  10     a  None    None   None  None            147   
2353                  10     a  None    None   None  None             47   
2354                  10     a  None    None   None  None             48   
2355                  10  None  None    None   None  None            532   

      favorite_count display_text_range  \
2351             111           [0, 120]   
2352             311           [0, 137]   
2353             128           [0, 130]   
2354             132           [0, 139]   
2355            2535           [0, 131]   

                                              jpg_url  img_num  \
2351  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg        1   
2352  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg        1   
2353  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg        1   
2354  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg        1   
2355  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg        1   

                          p1   p1_conf p1_dog                  p2   p2_conf  \
2351      miniature_pinscher  0.560311   True          Rottweiler  0.243682   
2352     Rhodesian_ridgeback  0.408143   True             redbone  0.360687   
2353         German_shepherd  0.596461   True            malinois  0.138584   
2354                 redbone  0.506826   True  miniature_pinscher  0.074192   
2355  Welsh_springer_spaniel  0.465074   True              collie  0.156665   

     p2_dog                   p3   p3_conf p3_dog  
2351   True             Doberman  0.154629   True  
2352   True   miniature_pinscher  0.222752   True  
2353   True           bloodhound  0.116197   True  
2354   True  Rhodesian_ridgeback  0.072010   True  
2355   True    Shetland_sheepdog  0.061428   True

In [325]:
df_main_cleaned['source'] = df_main_cleaned['source'].str.extract("\>(.*?)\<", expand = True)
df_main_cleaned.iloc[:,:3].tail(5)

tweet_id                 timestamp              source
2351  666049248165822465 2015-11-16 00:24:50+00:00  Twitter for iPhone
2352  666044226329800704 2015-11-16 00:04:52+00:00  Twitter for iPhone
2353  666033412701032449 2015-11-15 23:21:54+00:00  Twitter for iPhone
2354  666029285002620928 2015-11-15 23:05:30+00:00  Twitter for iPhone
2355  666020888022790149 2015-11-15 22:32:08+00:00  Twitter for iPhone

In [326]:
df_main_cleaned['source'].value_counts()

Twitter for iPhone    1932
Twitter Web Client      28
TweetDeck               11
Name: source, dtype: int64

Column is way more useable now!

**5. Split the text range into two separate columns - TIDINESS ISSUE!**

In [327]:
df_main_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1971 entries, 0 to 2355
Data columns (total 26 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            1971 non-null   object             
 1   timestamp           1971 non-null   datetime64[ns, UTC]
 2   source              1971 non-null   object             
 3   text                1971 non-null   object             
 4   expanded_urls       1971 non-null   object             
 5   rating_numerator    1971 non-null   int64              
 6   rating_denominator  1971 non-null   int64              
 7   name                1971 non-null   object             
 8   doggo               1971 non-null   object             
 9   floofer             1971 non-null   object             
 10  pupper              1971 non-null   object             
 11  puppo               1971 non-null   object             
 12  retweet_count       1971 non-null 

In [328]:
df_main_cleaned['display_text_range'].head(5)

0     [0, 85]
1    [0, 138]
2    [0, 121]
3     [0, 79]
4    [0, 138]
Name: display_text_range, dtype: object

In [329]:
df_main_cleaned['lower_text_range'] = df_main_cleaned['display_text_range'].apply(lambda x: x[0])

df_main_cleaned['upper_text_range'] = df_main_cleaned['display_text_range'].apply(lambda x: x[1])
df_main_cleaned.drop('display_text_range', axis = 1, inplace = True)

In [330]:
df_main_cleaned[["lower_text_range", "upper_text_range"]].head()

lower_text_range  upper_text_range
0                 0                85
1                 0               138
2                 0               121
3                 0                79
4                 0               138

As the display_text_range just contained lists we could extract them easily with list comprehension.

**6. Merging the different classification columns (doggo, floofer etc.) into one column and removing "None" entries - TIDINESS ISSUE!**

In [331]:
for dog_type in ['doggo', 'floofer', 'pupper', 'puppo']:
    df_main_cleaned[dog_type].replace('None', '', inplace = True)

df_main_cleaned['dog_type'] = df_main_cleaned['doggo'].map(str) + df_main_cleaned['floofer'].map(str) + df_main_cleaned['pupper'].map(str) + df_main_cleaned['puppo'].map(str)

In [332]:
df_main_cleaned['dog_type'].value_counts()

                1668
pupper           201
doggo             63
puppo             22
doggopupper        8
floofer            7
doggofloofer       1
doggopuppo         1
Name: dog_type, dtype: int64

In [333]:
df_main_cleaned['dog_type'].replace('', np.nan, inplace = True)

In [334]:
df_main_cleaned['dog_type'].value_counts() 

pupper          201
doggo            63
puppo            22
doggopupper       8
floofer           7
doggofloofer      1
doggopuppo        1
Name: dog_type, dtype: int64

In [335]:
df_main_cleaned.query('dog_type != ("pupper", "doggo", "puppo", "floofer") & ~dog_type.isnull()').count()

tweet_id              10
timestamp             10
source                10
text                  10
expanded_urls         10
rating_numerator      10
rating_denominator    10
name                  10
doggo                 10
floofer               10
pupper                10
puppo                 10
retweet_count         10
favorite_count        10
jpg_url               10
img_num               10
p1                    10
p1_conf               10
p1_dog                10
p2                    10
p2_conf               10
p2_dog                10
p3                    10
p3_conf               10
p3_dog                10
lower_text_range      10
upper_text_range      10
dog_type              10
dtype: int64

In [336]:
df_main_cleaned.drop(df_main_cleaned.query('dog_type != ("pupper", "doggo", "puppo", "floofer") & ~dog_type.isnull()').index, inplace=True)

In [337]:
df_main_cleaned.drop(['doggo', 'floofer', 'pupper', 'puppo',], inplace = True, axis = 1)

As there were some wrong types (multiple categories) I dropped those (9 entries) and the original columns, which are not necessary anymore.

**7. Remove incorrect names - QUALITY ISSUE!**

As the incorrect dog names are all lowercase, so I can look for lowercase names.

Instead of dropping those rows I've decided to rename those entries, because watching some tweets with incorrect names, we can see that those are indeed dogs (I looked at about 5).

The names are even provided in the tweet (for my sample of 5 at least), but they didn't get parsed correctly I guess.

In [338]:
df_main_cleaned[df_main_cleaned['name'].str.islower()]['expanded_urls'].values

array(['https://twitter.com/dog_rates/status/887517139158093824/video/1',
       'https://twitter.com/dog_rates/status/881536004380872706/video/1',
       'https://twitter.com/dog_rates/status/859196978902773760/video/1',
       'https://twitter.com/dog_rates/status/855459453768019968/photo/1,https://twitter.com/dog_rates/status/855459453768019968/photo/1',
       'https://twitter.com/dog_rates/status/828650029636317184/photo/1,https://twitter.com/dog_rates/status/828650029636317184/photo/1,https://twitter.com/dog_rates/status/828650029636317184/photo/1',
       'https://twitter.com/dog_rates/status/806219024703037440/photo/1',
       'https://twitter.com/dog_rates/status/792913359805018113/photo/1,https://twitter.com/dog_rates/status/792913359805018113/photo/1,https://twitter.com/dog_rates/status/792913359805018113/photo/1,https://twitter.com/dog_rates/status/792913359805018113/photo/1',
       'https://twitter.com/dog_rates/status/772581559778025472/photo/1,https://twitter.com/dog_ra

In [339]:
df_main_cleaned[df_main_cleaned['name'].str.islower()]['name'].value_counts()

a               55
the              7
an               6
very             4
one              4
quite            3
just             2
getting          2
my               1
this             1
incredibly       1
actually         1
all              1
officially       1
by               1
infuriating      1
light            1
such             1
unacceptable     1
not              1
his              1
space            1
Name: name, dtype: int64

In [340]:
#df_main_cleaned.drop(df_main_cleaned[df_main_cleaned['name'].str.islower()]['name'].index, inplace=True)

for index, row in df_main_cleaned[df_main_cleaned['name'].str.islower()].iterrows():
   df_main_cleaned.loc[index, 'name'] = 'None'

In [341]:
df_main_cleaned[df_main_cleaned['name'].str.islower()]['name'].value_counts()

Series([], Name: name, dtype: int64)

In [342]:
df_main_cleaned['name'].value_counts()

None       616
Charlie     11
Lucy        10
Oliver      10
Cooper      10
          ... 
Ivar         1
Eevee        1
Lili         1
Oreo         1
Baloo        1
Name: name, Length: 910, dtype: int64

In [343]:
df_main_cleaned.head()

tweet_id                 timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name  retweet_count  favorite_count  \
0                  10   Phineas           8853           39467   
1                  10     Tilly           6514           33819   
2                  10    Archie           4328           25461   
3                  10     Darla           8964           42908   
4                  10  Franklin           9774           41048   

                                           jpg_url  img_num           p1  \
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg        1       orange   
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg        1    Chihuahua   
2  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg        1    Chihuahua   
3  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg        1  paper_towel   
4  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg        2       basset   

    p1_conf p1_dog                  p2   p2_conf p2_dog  \
0  0.097049  False               bagel  0.085851  False   
1  0.323581   True            Pekinese  0.090647   True   
2  0.716012   True            malamute  0.078253   True   
3  0.170278  False  Labrador_retriever  0.168086   True   
4  0.555712   True    English_springer  0.225770   True   

                            p3   p3_conf p3_dog  lower_text_range  \
0                       banana  0.076110  False                 0   
1                     papillon  0.068957   True                 0   
2                       kelpie  0.031379   True                 0   
3                      spatula  0.040836  False                 0   
4  German_short-haired_pointer  0.175219   True                 0   

   upper_text_range dog_type  
0                85      NaN  
1               138      NaN  
2               121      NaN  
3                79      NaN  
4               138      NaN

In [344]:
df_main_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1961 entries, 0 to 2355
Data columns (total 24 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            1961 non-null   object             
 1   timestamp           1961 non-null   datetime64[ns, UTC]
 2   source              1961 non-null   object             
 3   text                1961 non-null   object             
 4   expanded_urls       1961 non-null   object             
 5   rating_numerator    1961 non-null   int64              
 6   rating_denominator  1961 non-null   int64              
 7   name                1961 non-null   object             
 8   retweet_count       1961 non-null   int64              
 9   favorite_count      1961 non-null   int64              
 10  jpg_url             1961 non-null   object             
 11  img_num             1961 non-null   int64              
 12  p1                  1961 non-null 

**8. Just use the column with the highest % and the respective prediction - TIDINESS ISSUE!**

In [345]:
df_main_cleaned.query("p2_conf > p1_conf")

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, retweet_count, favorite_count, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, lower_text_range, upper_text_range, dog_type]
Index: []

In [346]:
df_main_cleaned.query("p3_conf > p1_conf")

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, retweet_count, favorite_count, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, lower_text_range, upper_text_range, dog_type]
Index: []

In [347]:
df_main_cleaned.query("p3_conf > p2_conf")

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, retweet_count, favorite_count, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, lower_text_range, upper_text_range, dog_type]
Index: []

Checking the order of the columns - seems like it is ordered from p1 to p3. But the highest prediction % doesn't always mean that it's really a dog (3rd record has p1 as a paper_towel and p2 as a Labrador_retriever for example).

In [348]:
# This is not plagiarized, it is just a very common way to solve problems like this. Before I had it like this:
# https://stackoverflow.com/questions/48872234/using-apply-in-pandas-lambda-functions-with-multiple-if-statements
# That solution of course is very unclean and a quick google search led to this solution. This is not some niche solution nobody gets, it's very common on the internet.

def get_correct_breed(x):
    if x[0] == True:
        return x[1]
    elif x[2] == True:
        return x[3]
    else:
        return x[4]
    
df_main_cleaned["breed"] = df_main_cleaned[["p1_dog", "p1", "p2_dog", "p2", "p3"]].apply(get_correct_breed, axis = 1)
df_main_cleaned["conf"] = df_main_cleaned[["p1_dog", "p1_conf", "p2_dog", "p2_conf", "p3_conf"]].apply(get_correct_breed, axis = 1)

In [349]:
df_main_cleaned.iloc[:, 12:]

p1   p1_conf p1_dog                  p2   p2_conf  \
0                     orange  0.097049  False               bagel  0.085851   
1                  Chihuahua  0.323581   True            Pekinese  0.090647   
2                  Chihuahua  0.716012   True            malamute  0.078253   
3                paper_towel  0.170278  False  Labrador_retriever  0.168086   
4                     basset  0.555712   True    English_springer  0.225770   
...                      ...       ...    ...                 ...       ...   
2351      miniature_pinscher  0.560311   True          Rottweiler  0.243682   
2352     Rhodesian_ridgeback  0.408143   True             redbone  0.360687   
2353         German_shepherd  0.596461   True            malinois  0.138584   
2354                 redbone  0.506826   True  miniature_pinscher  0.074192   
2355  Welsh_springer_spaniel  0.465074   True              collie  0.156665   

     p2_dog                           p3   p3_conf p3_dog  lower_text_range  \
0     False                       banana  0.076110  False                 0   
1      True                     papillon  0.068957   True                 0   
2      True                       kelpie  0.031379   True                 0   
3      True                      spatula  0.040836  False                 0   
4      True  German_short-haired_pointer  0.175219   True                 0   
...     ...                          ...       ...    ...               ...   
2351   True                     Doberman  0.154629   True                 0   
2352   True           miniature_pinscher  0.222752   True                 0   
2353   True                   bloodhound  0.116197   True                 0   
2354   True          Rhodesian_ridgeback  0.072010   True                 0   
2355   True            Shetland_sheepdog  0.061428   True                 0   

      upper_text_range dog_type                   breed      conf  
0                   85      NaN                  banana  0.076110  
1                  138      NaN               Chihuahua  0.323581  
2                  121      NaN               Chihuahua  0.716012  
3                   79      NaN      Labrador_retriever  0.168086  
4                  138      NaN                  basset  0.555712  
...                ...      ...                     ...       ...  
2351               120      NaN      miniature_pinscher  0.560311  
2352               137      NaN     Rhodesian_ridgeback  0.408143  
2353               130      NaN         German_shepherd  0.596461  
2354               139      NaN                 redbone  0.506826  
2355               131      NaN  Welsh_springer_spaniel  0.465074  

[1961 rows x 14 columns]

In [350]:
df_main_cleaned.drop(df_main_cleaned.columns[12:21], inplace = True, axis = 1)

In [351]:
df_main_cleaned.head()

tweet_id                 timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name  retweet_count  favorite_count  \
0                  10   Phineas           8853           39467   
1                  10     Tilly           6514           33819   
2                  10    Archie           4328           25461   
3                  10     Darla           8964           42908   
4                  10  Franklin           9774           41048   

                                           jpg_url  img_num  lower_text_range  \
0  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg        1                 0   
1  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg        1                 0   
2  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg        1                 0   
3  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg        1                 0   
4  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg        2                 0   

   upper_text_range dog_type               breed      conf  
0                85      NaN              banana  0.076110  
1               138      NaN           Chihuahua  0.323581  
2               121      NaN           Chihuahua  0.716012  
3                79      NaN  Labrador_retriever  0.168086  
4               138      NaN              basset  0.555712

We've correctly extracted the prediction with the highest chance now (and of course checked if it's a dog).

**9. Standardize the breeds, because some are uppercase - QUALITY ISSUE!**

In [352]:
df_main_cleaned['breed'] = df_main_cleaned['breed'].str.lower()

In [353]:
df_main_cleaned['breed'].value_counts().tail(10)

boathouse      1
common_newt    1
chain          1
siamese_cat    1
plunger        1
titi           1
black_swan     1
grand_piano    1
hammerhead     1
seashore       1
Name: breed, dtype: int64

In [354]:
df_main_cleaned.head(20)

tweet_id                 timestamp              source  \
0   892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1   892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2   891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3   891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4   891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   
5   891087950875897856 2017-07-29 00:08:17+00:00  Twitter for iPhone   
6   890971913173991426 2017-07-28 16:27:12+00:00  Twitter for iPhone   
7   890729181411237888 2017-07-28 00:22:40+00:00  Twitter for iPhone   
8   890609185150312448 2017-07-27 16:25:51+00:00  Twitter for iPhone   
9   890240255349198849 2017-07-26 15:59:51+00:00  Twitter for iPhone   
10  890006608113172480 2017-07-26 00:31:25+00:00  Twitter for iPhone   
11  889880896479866881 2017-07-25 16:11:53+00:00  Twitter for iPhone   
12  889665388333682689 2017-07-25 01:55:32+00:00  Twitter for iPhone   
13  889638837579907072 2017-07-25 00:10:02+00:00  Twitter for iPhone   
14  889531135344209921 2017-07-24 17:02:04+00:00  Twitter for iPhone   
15  889278841981685760 2017-07-24 00:19:32+00:00  Twitter for iPhone   
16  888917238123831296 2017-07-23 00:22:39+00:00  Twitter for iPhone   
17  888804989199671297 2017-07-22 16:56:37+00:00  Twitter for iPhone   
18  888554962724278272 2017-07-22 00:23:06+00:00  Twitter for iPhone   
20  888078434458587136 2017-07-20 16:49:33+00:00  Twitter for iPhone   

                                                 text  \
0   This is Phineas. He's a mystical boy. Only eve...   
1   This is Tilly. She's just checking pup on you....   
2   This is Archie. He is a rare Norwegian Pouncin...   
3   This is Darla. She commenced a snooze mid meal...   
4   This is Franklin. He would like you to stop ca...   
5   Here we have a majestic great white breaching ...   
6   Meet Jax. He enjoys ice cream so much he gets ...   
7   When you watch your owner call another dog a g...   
8   This is Zoey. She doesn't want to be one of th...   
9   This is Cassie. She is a college pup. Studying...   
10  This is Koda. He is a South Australian decksha...   
11  This is Bruno. He is a service shark. Only get...   
12  Here's a puppo that seems to be on the fence a...   
13  This is Ted. He does his best. Sometimes that'...   
14  This is Stuart. He's sporting his favorite fan...   
15  This is Oliver. You're witnessing one of his m...   
16  This is Jim. He found a fren. Taught him how t...   
17  This is Zeke. He has a new stick. Very proud o...   
18  This is Ralphus. He's powering up. Attempting ...   
20  This is Gerald. He was just told he didn't get...   

                                        expanded_urls  rating_numerator  \
0   https://twitter.com/dog_rates/status/892420643...                13   
1   https://twitter.com/dog_rates/status/892177421...                13   
2   https://twitter.com/dog_rates/status/891815181...                12   
3   https://twitter.com/dog_rates/status/891689557...                13   
4   https://twitter.com/dog_rates/status/891327558...                12   
5   https://twitter.com/dog_rates/status/891087950...                13   
6   https://gofundme.com/ydvmve-surgery-for-jax,ht...                13   
7   https://twitter.com/dog_rates/status/890729181...                13   
8   https://twitter.com/dog_rates/status/890609185...                13   
9   https://twitter.com/dog_rates/status/890240255...                14   
10  https://twitter.com/dog_rates/status/890006608...                13   
11  https://twitter.com/dog_rates/status/889880896...                13   
12  https://twitter.com/dog_rates/status/889665388...                13   
13  https://twitter.com/dog_rates/status/889638837...                12   
14  https://twitter.com/dog_rates/status/889531135...                13   
15  https://twitter.com/dog_rates/status/889278841...                13   
16  https://twitter.com/dog_

In [355]:
df_main_cleaned.reset_index(drop = True, inplace = True)

In [356]:
df_main_cleaned.head(20)

tweet_id                 timestamp              source  \
0   892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1   892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2   891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3   891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4   891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   
5   891087950875897856 2017-07-29 00:08:17+00:00  Twitter for iPhone   
6   890971913173991426 2017-07-28 16:27:12+00:00  Twitter for iPhone   
7   890729181411237888 2017-07-28 00:22:40+00:00  Twitter for iPhone   
8   890609185150312448 2017-07-27 16:25:51+00:00  Twitter for iPhone   
9   890240255349198849 2017-07-26 15:59:51+00:00  Twitter for iPhone   
10  890006608113172480 2017-07-26 00:31:25+00:00  Twitter for iPhone   
11  889880896479866881 2017-07-25 16:11:53+00:00  Twitter for iPhone   
12  889665388333682689 2017-07-25 01:55:32+00:00  Twitter for iPhone   
13  889638837579907072 2017-07-25 00:10:02+00:00  Twitter for iPhone   
14  889531135344209921 2017-07-24 17:02:04+00:00  Twitter for iPhone   
15  889278841981685760 2017-07-24 00:19:32+00:00  Twitter for iPhone   
16  888917238123831296 2017-07-23 00:22:39+00:00  Twitter for iPhone   
17  888804989199671297 2017-07-22 16:56:37+00:00  Twitter for iPhone   
18  888554962724278272 2017-07-22 00:23:06+00:00  Twitter for iPhone   
19  888078434458587136 2017-07-20 16:49:33+00:00  Twitter for iPhone   

                                                 text  \
0   This is Phineas. He's a mystical boy. Only eve...   
1   This is Tilly. She's just checking pup on you....   
2   This is Archie. He is a rare Norwegian Pouncin...   
3   This is Darla. She commenced a snooze mid meal...   
4   This is Franklin. He would like you to stop ca...   
5   Here we have a majestic great white breaching ...   
6   Meet Jax. He enjoys ice cream so much he gets ...   
7   When you watch your owner call another dog a g...   
8   This is Zoey. She doesn't want to be one of th...   
9   This is Cassie. She is a college pup. Studying...   
10  This is Koda. He is a South Australian decksha...   
11  This is Bruno. He is a service shark. Only get...   
12  Here's a puppo that seems to be on the fence a...   
13  This is Ted. He does his best. Sometimes that'...   
14  This is Stuart. He's sporting his favorite fan...   
15  This is Oliver. You're witnessing one of his m...   
16  This is Jim. He found a fren. Taught him how t...   
17  This is Zeke. He has a new stick. Very proud o...   
18  This is Ralphus. He's powering up. Attempting ...   
19  This is Gerald. He was just told he didn't get...   

                                        expanded_urls  rating_numerator  \
0   https://twitter.com/dog_rates/status/892420643...                13   
1   https://twitter.com/dog_rates/status/892177421...                13   
2   https://twitter.com/dog_rates/status/891815181...                12   
3   https://twitter.com/dog_rates/status/891689557...                13   
4   https://twitter.com/dog_rates/status/891327558...                12   
5   https://twitter.com/dog_rates/status/891087950...                13   
6   https://gofundme.com/ydvmve-surgery-for-jax,ht...                13   
7   https://twitter.com/dog_rates/status/890729181...                13   
8   https://twitter.com/dog_rates/status/890609185...                13   
9   https://twitter.com/dog_rates/status/890240255...                14   
10  https://twitter.com/dog_rates/status/890006608...                13   
11  https://twitter.com/dog_rates/status/889880896...                13   
12  https://twitter.com/dog_rates/status/889665388...                13   
13  https://twitter.com/dog_rates/status/889638837...                12   
14  https://twitter.com/dog_rates/status/889531135...                13   
15  https://twitter.com/dog_rates/status/889278841...                13   
16  https://twitter.com/dog_

We've made the breeds all lowercase and additionally reset the index, as we've dropped some rows along the way and the index doesn't represent the correct amount of rows anymore.

In [357]:
df_main_cleaned.to_csv('twitter_archive_master.csv', index = False)